src/System.Linq.Queryable/src/System/Linq/Queryable.cs

```diff
@@ -842,6 +842,23 @@ public static TSource SingleOrDefault<TSource>(this IQueryable<TSource> source,
                     ));
         }
 
+        public static bool TrySingle<TSource>(this IQueryable<TSource> source, out TSource element)
+        {
+            Tuple<bool, TSource> result = TrySingle(source);
+            element = result.Item2;
+            return result.Item1;
+        }
+
+        internal static Tuple<bool, TSource> TrySingle<TSource>(this IQueryable<TSource> source)
+        {
+            if (source == null)
+                throw Error.ArgumentNull(nameof(source));
+            return source.Provider.Execute<Tuple<bool, TSource>>(
+                Expression.Call(
+                    null,
+                    CachedReflectionInfo.TrySingle_TSource_1(typeof(TSource)), source.Expression));
+        }
+
```

Comment:
@bartdesmet thanks for binging this to my attention. I have provided my general feedback on the feature at https://github.com/dotnet/corefx/issues/6073#issuecomment-266867512. 

Besides that, we seem to share many of the same experience trying to overcome the limitations of the `IQueryable`/`IQueryProvider` story, e.g. how the ability to compose expressions for deferred execution is conflated with types representing collections. But this matter seems to go way beyond what this PR is about so I believe it would be great if we could have a separate discussion on it one of these days.